apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -21,11 +29,76 @@ def get_file_names_as_array(raw_file_names: str) -> list[str]:
     return next(reader)
 
 
+# Split optional file names into two categories: wildcard patterns and plain file names
+# For the wildcards, convert into regular expressions and return a two-item list with both the regex and a plain display version
+# For the plain names, remove escapes for displaying and return in a separate list
+def extract_patterns(optional_files: list[str]) -> tuple[list[list[str]], list[str]]:
+    # Try converting all file names to regexes; if no wildcard is present, glob_to_regex returns an empty string
+    optional_file_regex_raw = [glob_to_regex(x) for x in optional_files]
```

Comment:
I think this function can be written using a single for loop and appending to the `optional_file_plain` and `optional_file_regex` lists before returning them. As currently written, I think this uses too many comprehensions and is hard to read. This seems like a case where the logic is too complex to be handled cleanly with list comprehensions.

Reply:
You're right. After rewriting as a loop, the logic is simpler than I expected and much easier to read.